sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/FailureSafeParser.scala

```diff
@@ -55,11 +56,15 @@ class FailureSafeParser[IN](
 
   def parse(input: IN): Iterator[InternalRow] = {
     try {
-     if (skipParsing) {
-       Iterator.single(InternalRow.empty)
-     } else {
-       rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
-     }
+      if (skipParsing) {
+        if (unparsedRecordIsNonEmpty(input)) {
```

Comment:
cc @cloud-fan and @gatorsmile, if you don't mind, I would like to revert #21909. WDYT?

Reply:
that's a good catch! I think the idea of count optimization still makes sense, but our parser is so flexible and there are many corner cases we need to think of.

cc @MaxGekk , how hard do you think it is to fix it? If it's too hard, +1 to revert it. I think a safer approach maybe, only enable this count optimization under some certain cases that are 100% safe. (whitelist).

Reply:
Yea, that's safer approach if possible but to do that, we should manually check the input like the current PR. It doesn't looks a good idea to me to check input outside of `JacksonParser`.

The problem is, we cannot distinguish the cases below without parsing:

```
[{...}, {...}]
```

```
[]
```

```
{...}
```

```
# empty string
```

One line (`input: IN`) can be, 0 record, 1 record and multiple records.

Reply:
> The case when an user sets StructType for arrays, can be excluded from the count optimization in advance.

How are you going to exclude this without checking the input? The input decides the number of records to return from the input `IN`.

> Regarding empty (blank) string, before #23543 they are considered as bad records (appear in results). And count() produced pretty consistent results. 

I think this was mistake. It had to be reverted and it was reverted. Empty string isn't a JSON.

> I think we should answer to more generic question - which input conform to empty `StructType()`.

I basically think an empty `StructType` basically means empty object `{}`. I was thinking it's not controversial.


> Till we answer to the above question, reverting of the #21909 just move us from one "bad" behavior to one another "bad" behavior.

No, it will moves one "bad" behaviour to the previous behaviour that keeps backward compatibility.

Reply:
Also, the counting behaviour here looks definitely wrong and counter-intuitive since it doesn't match to `show()` obviously. It's something we should definitely fix, or revert the change that caused it.

If previous behaviour isn't coherent at all, yes, we should change but if it makes sense, we better stick to it and we can consider that's "right" behaviour.

Also, let's make the behavioural changes clear in other PRs next time.

Reply:
> I basically think an empty StructType basically means empty object {}. I was thinking it's not controversial.

I mean when required schema is `an empty StructType`. If you accept only `{}` as valid input for `count()`, you will get absolutely wrong results.

Reply:
> I think this was mistake. It had to be reverted and it was reverted. Empty string isn't a JSON.

The changes was not about `string isn't a JSON`. The reverted PR introduced 2 types of input - `bad` records and valid records. By reverting it, we introduced 3rd category - hidden (or invisible for users) input.

Reply:
> No, it will moves one "bad" behaviour to the previous behaviour that keeps backward compatibility.

Backward compatibility to what? The behavior was released in 2.4

Reply:
> Backward compatibility to what? The behavior was released in 2.4

That was a mistake to fix, @MaxGekk

Reply:
I know what you're arguing about conversion. The conversion makes the count different, I know. I think I stood against at that time. We can leave that out here.

Let's match to:

```
[{...}, {...}] => 2
[] => 0
{...} => 1
# empty string => 0
```

I don't see that we should fix it to:

```
[{...}, {...}] => 1
[] => 1
{...} => 1
# empty string => 1
```

where the output is:

```
+---+
|  a|
+---+
|  1|
|  2|
+---+
```

This fix was a mistake

Reply:
Let's go back to the original one, and try it again. People complain about the behaviour changes like this JIRA.

Reply:
I agree with that our JSON / CSV have some holes about parsing, counting, column pruning (in particular when it's malformed), etc. I was thinking about whitelisting the problems and behaviours at 3.0.0.

I don't think this behaviour was noticed to reviewers when #21909 was merged at that time while this is pretty critical to discuss as you said above. Let's do this again probably with whitelisting behaviours clearly. I think we should avoid this forth and back with the doc.

Reply:
ok. What about just put the count optimization under a separate flag? The optimization makes a lot of sense for most of users ( > 90% from my point of view of course) who has huge well formatted JSON files.

Reply:
Honestly, I don't quite like the bunch or configurations we have. Can we instead use text source to count them for now?

Reply:
> we introduced 3rd category - hidden (or invisible for users) input

This is a good definition. So there are 3 kinds of records: valid, bad and hidden.

The definition of bad records is query-dependent. If the corrected column is not read, then this is still a valid record.

`df.show()` displays valid records, and bad records if the mode is permissive. This means the result of `df.show()` depends on a config. If we want to make `df.count()` consistent with `df.show()`, maybe we should let it depend on the same config?

Reply:
It should be good to define the behaviour but I don't think https://github.com/apache/spark/pull/23665#discussion_r251276720 behaviour makes sense to users. It will end up with text datasource + count().

Reply:
Like, how are we going to explain this to users? When you count, the record is separated by newlines but other operations recognise JSON array and empty strings .. , and count is a special case? CSV doesn't do that fwiw - it filters empty strings. I think we shouldn't do reasoning in particular for such basic operations.

If I were a user, I would think it's counter initiative. Is it me alone thinking in this way? This case specifically doesn't look straightforward. I know there are issues about parsing stuff and we should define the behaviour but this case looks not.

Reply:
> Like, how are we going to explain this to users?

Outside of datasources, `count()` has very well defined semantic - number of rows matched to required schema. In the case of `count()`, the required schema is `Struct(Array())`. From user's view, `count()` doesn't require any field presented in counted row.

If you would like to see the same number of rows in `show()` output and what `count()` returns, you need to push full datasource schema as required schema otherwise there are always malformed input on which you will see different results.

Reply:
> otherwise there are always malformed input on which you will see different results

I think for permissive mode, the results(at least the counts) are always same even if some input are malformed? Otherwise, it seems like users only want to count the number of lines, and they should read the json files as text and do count.

Reply:
re: https://github.com/apache/spark/pull/23665#discussion_r251508679

This is reasoning about the count. (see below, continues)

> I think for permissive mode, the results(at least the counts) are always same even if some input are malformed? 

To be 100% about the correct results, we should always parse everything although we're doing the current way for optimization and it started to have some inconsistent results.

Yes, so we don't convert 100%. In that case, we should at least parse `StructType()` which I guess empty object `{...}`. I think that's what JSON did before the pointed PRs above.

Reply:
> It looks to me that we can enable the count optimization only for {...}, and fallback to parsing for other cases. @MaxGekk do you think this is applicable? If it's a simple fix, let's do it for branch 2.4 as well, otherwise +1 for reverting it from 2.4 and re-do it at master.

We can but we should add a if-else for each input + checking some characters since it targets to avoid to parse.